guide/src/migration.md

```diff
@@ -0,0 +1,183 @@
+# Appendix B: Migration Guides for major version changes
+
+## from 0.8.* to 0.9
+
+### `#[new]` interface
+[`PyRawObject`](https://docs.rs/pyo3/0.8.5/pyo3/type_object/struct.PyRawObject.html)
+is now removed and our syntax for constructor changed.
+
+Before:
+```compile_fail
+#[pyclass]
+struct MyClass {}
+
+#[pymethods]
+impl MyClass {
+   #[new]
+   fn new(obj: &PyRawObject) {
+       obj.init(MyClass { })
+   }
+}
+```
+
+After:
+```
+# use pyo3::prelude::*;
+#[pyclass]
+struct MyClass {}
+
+#[pymethods]
+impl MyClass {
+   #[new]
+   fn new() -> Self {
+       MyClass {}
+   }
+}
+```
+
+Basically you can return `Self` or `Result<Self>` directly.
+For more, see [the constructor section](https://pyo3.rs/master/class.html#constructor) of this guide.
+
+### PyCell
+PyO3 0.9 introduces [`PyCell`](https://pyo3.rs/master/doc/pyo3/pycell/struct.PyCell.html), which is
+a [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html) like object wrapper
+for dynamically ensuring
+[Rust's rule of Reference](https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html#the-rules-of-references).
+
+For `#[pymethods]` or `#[pyfunction]`s, `PyCell` works without any change.
+Just throw errors when there happens invalid borrowing.
+
+Here is an example.
+```
+# use pyo3::prelude::*;
+#[pyclass]
+struct Names {
+    names: Vec<String>
+}
+
+#[pymethods]
+impl Names {
+   #[new]
+   fn new() -> Self {
+       Names { names: vec![] }
+   }
+   fn merge(&mut self, other: &mut Names) {
+       self.names.append(&mut other.names)
+   }
+}
+# let gil = Python::acquire_gil();
+# let py = gil.python();
+# let names = PyCell::new(py, Names::new()).unwrap();
+# let borrow_mut_err = py.get_type::<pyo3::pycell::PyBorrowMutError>();
+# pyo3::py_run!(py, names borrow_mut_err, r"
+# try:
+#    names.merge(names)
+#    assert False, 'Unreachable'
+# except Exception as e:
+#    isinstance(e, borrow_mut_err)
+# ");
+```
+`Names` has `merge` method, which takes `&mut self` and `&mut Self`.
+Given this `#[pyclass]`, calling `names.merge(names)` in Python raises `PyBorrowMutError` exception,
+since it requires two mutable borrows of `names`,
+
+However, for `#[pyproto]` and some functions, you need to manually fix codes.
```

Comment:
We should probably link to the `#[pyproto]` section. Or move the `#[pyproto]` section here.